include/seqan3/argument_parser/validators.hpp

```diff
@@ -297,68 +290,347 @@ class file_ext_validator
+     */
+    bool path_is_readable(std::filesystem::path const & path) const
+    {
+        // Check if input directory is a readable.
+        if (std::filesystem::is_directory(path))
+        {
+            std::error_code ec{};
+            std::filesystem::directory_iterator{path, ec};  // if directory iterator cannot be created, ec will be set.
+            return !static_cast<bool>(ec); // test the error code.
+        }
+
+        // Must be a regular file.
+        if (!std::filesystem::is_regular_file(path))
+            throw std::invalid_argument{detail::to_string("Expected a regular file!", path, "!")};
+
+        std::ifstream file{path};
+        return file.is_open() && file.good();
+    }
+
+    /*!\brief Checks if the given path is writable.
+     * \param path The path to check.
+     * \returns `true` if writable, otherwise `false`.
+     */
+    bool path_is_writable(std::filesystem::path const & path) const
+    {
+        std::ofstream file{path};
+        bool is_open = file.is_open();
+        bool is_good = file.good();
+        file.close();
+
+        // Try to remove file.
+        std::filesystem::remove(path);
+
+        return is_open && is_good;
+    }
+
+    //!\brief Stores the extensions.
+    std::vector<std::string> extensions{};
+};
+
+/*!\brief A validator that checks if a given path is a valid input file.
+ * \ingroup argument_parser
+ * \implements seqan3::validator_concept
+ *
+ * \details
+ *
+ * On construction, the validator can receive a list (std::vector over std::string) of valid file extensions.
+ * The struct acts as a functor that throws a seqan3::parser_invalid_argument exception whenever a given filename's
+ * extension (sts::filesystem::path) is not in the given list of valid file extensions, if the file does not exist, or
+ * if the file does not have the proper read permissions.
+ *
+ * \snippet test/snippet/argument_parser/validators_input_file.cpp usage
+ *
+ * \note The validator works on every type that can be implicitly converted to std::filesystem::path.
+ */
+class input_file_validator : public file_validator_base
+{
+public:
+    //!\brief Type of values that are tested by the validator.
+    using file_validator_base::value_type;
+
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    input_file_validator() = default;                                         //!< Defaulted.
+    input_file_validator(input_file_validator const &) = default;             //!< Defaulted.
+    input_file_validator(input_file_validator &&) = default;                  //!< Defaulted.
+    input_file_validator & operator=(input_file_validator const &) = default; //!< Defaulted.
+    input_file_validator & operator=(input_file_validator &&) = default;      //!< Defaulted.
+    ~input_file_validator() = default;                                        //!< Defaulted.
+
+    //!\brief Import constructor for setting valid extensions.
```

Comment:
I think using declarations must not to be documented because the documentation is copied and an extra brief might confuse doxygen. (at least it was like that with member typedefs when I last checked)

Reply:
Mhmm, AFAIK doxygen ignores this completely. But you never know what the next doxygen version brings, so I documented it anyway. I will at least keep it as a comment so less experienced people can understand what is happening.

Reply:
yes it ignores this but I had trouble with typedefs where an extra brief actually introduced spurious documentation somewhere else. so yes a normal comment would be good